lib/LanguageServerRename/Model/Renamer.php

```diff
@@ -0,0 +1,314 @@
+                    count($locations),
+                    $count,
+                    number_format(microtime(true) - $start, 2),
+                    $this->timeoutSeconds
+                ));
+                return $this->locationsToWorkspaceEdit($locations, $oldName, $newName);
+            }
+
+            // I can't see how I can yield here.
+            // if ($count++ % 10) {
+            //     // give other co-routines a chance
+            //     yield new Delayed(0);
+            // }
+        }
+        
+        $this->clientApi->window()->showMessage()->info(sprintf(
+            'Found %s reference(s) to be renamed.',
+            count($locations)
+        ));
+
+        return $this->locationsToWorkspaceEdit($locations, $oldName, $newName);
+    }
+
+    private function documentAndPositionToNodeAndOffset(TextDocumentItem $textDocument, Position $position): array
+    {
+        $offset = PositionConverter::positionToByteOffset($position, $textDocument->text);
+        
+        $rootNode = $this->parser->parseSourceFile($textDocument->text);
+        $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
+        return [$offset, $node];
+    }
+
+    private function locationsToWorkspaceEdit(array $locations, string $oldName, string $newName): WorkspaceEdit
+    {
+        // group locations by document
+        $locationsByUri = [];
+        foreach ($locations as $location) {
+            /** @var Location $location */
+            $uri = (string)$location->uri();
+            if (!isset($locationsByUri[$uri])) {
+                $locationsByUri[$uri] = [];
+            }
+            $locationsByUri[$uri][] = $location;
+        }
+
+        $documentEdits = [];
+        foreach ($locationsByUri as $uri => $locations) {
+            list($textEdits, $renameOperation) = $this->documentLocationsToTextEdits($uri, $locations, $oldName, $newName);
+            $documentEdits[] = new TextDocumentEdit(
+                new VersionedTextDocumentIdentifier($uri, $this->getDocumentVersion($uri)),
+                $textEdits
+            );
+            if($renameOperation !== null)
+                $documentEdits[] = $renameOperation;
+        }
+
+        return new WorkspaceEdit(null, $documentEdits);
+    }
+
+    private function documentLocationsToTextEdits(string $documentUri, array $locations, string $oldName, string $newName): array
+    {
+        $edits = [];
+        $rename = null;
+        $documentContent = $this->getDocumentText($documentUri);
+        $rootNode = $this->parser->parseSourceFile($documentContent);
+        foreach ($locations as $location) {
+            /** @var Location $location */
+            $node = $rootNode->getDescendantNodeAtPosition($location->offset()->toInt());
+            
+            if(($r = $this->renameFileIfNeeded($node, $documentUri, $oldName, $newName)) !== null){
```

Comment:
would be happy if we removed the file-rename functionality for now, this is a more complex topic I think. Let's concentrate on a smaller set of features.

Reply:
What do you mean, 'more complex'? Specifically renaming when renaming the class itself is pretty straight-forward - either they have the same name or they don't (unless I am missing something). The other way around is the complicated bit, I think - when the class file is being moved..

Reply:
I mean it will spiral out of control probably. If we focus on a single feature then we can be happy with it faster and merge it, then see about other features.

Reply:
I am inclined to agree, although I'd be happy to hear the scenario you imagine.

Reply:
Off the top of my head the class moving stuff:

- Rename is moving.
- Moving is moving a file - finding out where that file should live depending on the namespace (Phpactor already has a solution for this problem)
- The current Phpactor implementation moves the file in GiT too (via. the source code filesystem)
- Dealing with class / trait / blah / aliases is always fun.
- The current RPC handler for move will also move any related test files.

There will always be more complexity that will take you by surprise.
-